x/upgrade/module.go

```diff
@@ -0,0 +1,96 @@
+
+	"github.com/cosmos/cosmos-sdk/client"
+	"github.com/cosmos/cosmos-sdk/codec"
+	codectypes "github.com/cosmos/cosmos-sdk/codec/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/module"
+	"github.com/cosmos/cosmos-sdk/x/upgrade"
+	cli "github.com/cosmos/cosmos-sdk/x/upgrade/client/cli"
+	upgradekeeper "github.com/cosmos/cosmos-sdk/x/upgrade/keeper"
+	upgradetypes "github.com/cosmos/cosmos-sdk/x/upgrade/types"
+	"github.com/grpc-ecosystem/grpc-gateway/runtime"
+	"github.com/spf13/cobra"
+	abci "github.com/tendermint/tendermint/abci/types"
+)
+
+var _ module.AppModule = AppModule{}
+
+const (
+	consensusVersion uint64 = 1
+)
+
+var (
+	_ module.AppModule      = AppModule{}
+	_ module.AppModuleBasic = AppModuleBasic{}
+)
+
+// AppModuleBasic implements the sdk.AppModuleBasic interface
+type AppModuleBasic struct {
+	upgrade.AppModuleBasic
+}
+
+// Name returns the ModuleName
+func (AppModuleBasic) Name() string {
+	return upgradetypes.ModuleName
+}
+
+// RegisterLegacyAminoCodec registers the upgrade types on the LegacyAmino codec
+func (AppModuleBasic) RegisterLegacyAminoCodec(cdc *codec.LegacyAmino) {
+	upgradetypes.RegisterLegacyAminoCodec(cdc)
+}
+
+// RegisterGRPCGatewayRoutes registers the gRPC Gateway routes for the upgrade module.
+func (AppModuleBasic) RegisterGRPCGatewayRoutes(clientCtx client.Context, mux *runtime.ServeMux) {
+	if err := upgradetypes.RegisterQueryHandlerClient(context.Background(), mux, upgradetypes.NewQueryClient(clientCtx)); err != nil {
+		panic(err)
+	}
+}
+
+// GetQueryCmd returns the cli query commands for this module
+func (AppModuleBasic) GetQueryCmd() *cobra.Command {
+	return GetQueryCmd()
+}
+
+// GetTxCmd returns the transaction commands for this module
+func (AppModuleBasic) GetTxCmd() *cobra.Command {
+	return cli.GetTxCmd()
+}
+
+func (b AppModuleBasic) RegisterInterfaces(registry codectypes.InterfaceRegistry) {
+	upgradetypes.RegisterInterfaces(registry)
+}
+
+// AppModule implements the sdk.AppModule interface by wrapping the standard
+// upgrade module and overwriting methods specifc to the msg server.
+type AppModule struct {
+	upgrade.AppModule
+}
+
+// NewAppModule returns the AppModule for the application's upgrade module.
+func NewAppModule(keeper upgradekeeper.Keeper) AppModule {
+	return AppModule{
+		upgrade.NewAppModule(keeper),
+	}
+}
+
+// RegisterInterfaces overwrites the default upgrade module's method with a
+// noop.
+func (AppModule) RegisterInterfaces(registry codectypes.InterfaceRegistry) {
+}
+
+// RegisterServices overwrites the default upgrade module's method with a
+// noop.
+func (am AppModule) RegisterServices(cfg module.Configurator) {
+}
+
+// ConsensusVersion implements AppModule/ConsensusVersion.
+func (AppModule) ConsensusVersion() uint64 { return consensusVersion }
+
+// BeginBlock overwrites the default upgrade module's BeginBlock method with a
+// noop.
+func (am AppModule) BeginBlock(ctx sdk.Context, req abci.RequestBeginBlock) {
+}
```

Comment:
Here's the most important state machine changes. By wrapping the upgrade module with our own and then overwriting these methods, we're removing any of the external functionality that's in the msg server/begin block, while preserving keeper access to the upgrade module to maintain 100% functionality with the IBC module.

Reply:
Don't we need to also overwrite `endBlock`?

Reply:
I thought the same, but I don't think the upgrade module has an endBlock

Reply:
my bad, confused it with `gov` module :D

Reply:
> Could we just keep the keeper and not the module?

this is a good idea! I think this would be the simplest approach, but I think we still need some way to register some of the upgrade types.

for example when the governance can [upgrade an IBC clients](https://github.com/celestiaorg/celestia-app/blob/c5ae8ac2d7f2d290f336faf170c7547bfd0929a2/app/app.go#L358)
```go
// NewSoftwareUpgradeProposalHandler creates a governance handler to manage new proposal types.
// It enables SoftwareUpgradeProposal to propose an Upgrade, and CancelSoftwareUpgradeProposal
// to abort a previously voted upgrade.
//
//nolint:staticcheck // we are intentionally using a deprecated proposal here.
func NewSoftwareUpgradeProposalHandler(k *keeper.Keeper) govtypes.Handler {
	return func(ctx sdk.Context, content govtypes.Content) error {
		switch c := content.(type) {
		case *types.SoftwareUpgradeProposal: <--
			return handleSoftwareUpgradeProposal(ctx, k, c)

		case *types.CancelSoftwareUpgradeProposal: <--
			return handleCancelSoftwareUpgradeProposal(ctx, k, c)

		default:
			return sdkerrors.Wrapf(sdkerrors.ErrUnknownRequest, "unrecognized software upgrade proposal content type: %T", c)
		}
	}
}
```